Firestore/core/test/firebase/firestore/local/remote_document_cache_test.mm

```diff
@@ -0,0 +1,222 @@
+using model::NoDocument;
+using model::OptionalMaybeDocumentMap;
+
+using testing::Matches;
+using testing::IsSupersetOf;
+using testing::UnorderedElementsAreArray;
+using testutil::DeletedDoc;
+using testutil::Doc;
+using testutil::Map;
+using testutil::Query;
+
+const char* kDocPath = "a/b";
+const char* kLongDocPath = "a/b/c/d/e/f";
+const int kVersion = 42;
+FieldValue::Map kDocData;
+
+template <typename MapType>
+std::vector<MaybeDocument> ActualDocuments(const MapType& docs) {
+  std::vector<MaybeDocument> result;
+  for (const auto& kv : docs) {
+    const absl::optional<MaybeDocument>& doc = kv.second;
+    if (doc) {
+      result.push_back(*doc);
+    }
+  }
+  return result;
+}
+
+MATCHER_P(HasExactlyDocs,
+          expected,
+          negation ? "missing docs" : "has exactly docs") {
+  std::vector<MaybeDocument> arg_docs = ActualDocuments(arg);
+  return testing::Value(arg_docs, UnorderedElementsAreArray(expected));
+}
+
+MATCHER_P(HasAtLeastDocs,
+          expected,
+          negation ? "missing docs" : "has at least docs") {
+  std::vector<MaybeDocument> arg_docs = ActualDocuments(arg);
+  return testing::Value(arg_docs, IsSupersetOf(expected));
+}
+
+}  // namespace
+
+RemoteDocumentCacheTest::RemoteDocumentCacheTest()
+    : persistence_{GetParam()()},
+      cache_{persistence_->remote_document_cache()} {
+  // essentially a constant, but can't be a compile-time one.
+  kDocData = Map("a", 1, "b", 2);
+}
+
+TEST_P(RemoteDocumentCacheTest, ReadDocumentNotInCache) {
+  persistence_->Run("test_read_document_not_in_cache", [&] {
+    ASSERT_EQ(absl::nullopt, cache_->Get(testutil::Key(kDocPath)));
+  });
+}
+
+TEST_P(RemoteDocumentCacheTest, SetAndReadADocument) {
+  SetAndReadTestDocument(kDocPath);
+}
+
+TEST_P(RemoteDocumentCacheTest, SetAndReadSeveralDocuments) {
+  persistence_->Run("test_set_and_read_several_documents", [=] {
+    std::vector<Document> written = {
+        SetTestDocument(kDocPath),
+        SetTestDocument(kLongDocPath),
+    };
+    OptionalMaybeDocumentMap read = cache_->GetAll(
+        DocumentKeySet{testutil::Key(kDocPath), testutil::Key(kLongDocPath)});
+    EXPECT_THAT(read, HasExactlyDocs(written));
+  });
+}
+
+TEST_P(RemoteDocumentCacheTest,
+       SetAndReadSeveralDocumentsIncludingMissingDocument) {
+  persistence_->Run(
+      "test_set_and_read_several_documents_including_missing_document", [=] {
+        std::vector<Document> written = {
+            SetTestDocument(kDocPath),
+            SetTestDocument(kLongDocPath),
+        };
+        OptionalMaybeDocumentMap read = cache_->GetAll(DocumentKeySet{
+            testutil::Key(kDocPath),
+            testutil::Key(kLongDocPath),
+            testutil::Key("foo/nonexistent"),
+        });
+        EXPECT_THAT(read, HasAtLeastDocs(written));
+        auto found = read.find(DocumentKey::FromPathString("foo/nonexistent"));
+        ASSERT_TRUE(found != read.end());
+        ASSERT_EQ(absl::nullopt, found->second);
```

Comment:
This seems to match what `HasExactlyDocs` does. Since it is also the only caller of HasAtLeastDocs we could consider using HasExactlyDocs and dropping HasAtLeastDocs.

Reply:
I don't think this is true.

This is checking that `read` includes a document entry for `"foo/nonexistent"` and that its value is `optional::none`. This is something that `HasExactlyDocs` and `HasAtLeastDocs` ignore.

Reply:
Got it, I misread `ASSERT_TRUE(found != read.end())` and thought it indicated the opposite.